pgxn/neon/pagestore_smgr.c

```diff
@@ -126,35 +128,338 @@ static UnloggedBuildPhase unlogged_build_phase = UNLOGGED_BUILD_NOT_IN_PROGRESS;
  * all prefetch responses has to be consumed.
  */
 
-#define MAX_PREFETCH_REQUESTS 128
+BufferTag	prefetch_requests[READ_BUFFER_SIZE];
+
+PrefetchState *MyPState;
 
-BufferTag	prefetch_requests[MAX_PREFETCH_REQUESTS];
-BufferTag	prefetch_responses[MAX_PREFETCH_REQUESTS];
-int			n_prefetch_requests;
-int			n_prefetch_responses;
-int			n_prefetched_buffers;
 int			n_prefetch_hits;
 int			n_prefetch_misses;
+int			n_prefetch_dupes;
+
 XLogRecPtr	prefetch_lsn;
 
+static void consume_prefetch_responses(void);
+static void prefetch_ensure_unused_space(void);
+static uint64 prefetch_register_buffer(BufferTag tag);
+static inline void prefetch_read(PrefetchRequest *entry);
+static inline void prefetch_request(PrefetchRequest *slot);
+static inline void prefetch_wait_for(uint64 ring_index);
+static inline void prefetch_cleanup(void);
+static void prefetch_move_buffers(uint64 to, uint64 from, int count);
+
+static XLogRecPtr neon_get_request_lsn(bool *latest, RelFileNode rnode,
+									   ForkNumber forknum, BlockNumber blkno);
+
+/*
+ * Make sure that there are no responses still in the buffer.
+ */
 static void
 consume_prefetch_responses(void)
 {
-	for (int i = n_prefetched_buffers; i < n_prefetch_responses; i++)
+	MemoryContext old = MemoryContextSwitchTo(MyPState->context);
```

Comment:
Took me a while to find where we do allocations where this takes effect. It's in `prefetch_read`, it calls `page_server->receive()`, which returns a palloc'd struct.

I'd recommend moving this `MemoryContextSwitchTo` into `prefetch_read` itself. As it is, it's a subtle requirement that the caller must've have done that. That's easy to miss if this gets refactored around in the future, and it could appear to work at first if the calling memory context is fairly long-lived.